### PR TITLE
Patch encoder/decoder to convert ABI addresses

### LINKED ISSUE #127: Patch encoder/decoder to convert ABI addresses
In https://github.com/vyperlang/titanoboa/pull/115 we introduced some [utilities](https://github.com/DanielSchiavini/titanoboa/blob/6f3fea7744570f8e657f4854c0bfb229ec940443/boa/contracts/utils.py) to convert ABI addresses into `Address` objects.
It would be better to change the eth-stdlib encoder and decoder to do that. That should work better with deeply nested objects.

### PR DESCRIPTION
**Pull request by DanielSchiavini**
Closes #127

### What I did
Got rid of boa.contracts.utils and applied the address conversion to all abi operations.

### How I did it
- Add a new test to make sure the previous implementation did not work recursively
- Inherit eth.codecs.abi classes to convert addresses
- Create an AddressWrapper that works with `vyper_object`

### How to verify it
- Existing tests should not be broken
- Test is updated to reproduce previous issue

### Description for the changelog
- Address strings are always converted to objects when using the ABI

### Cute Animal Picture
image

### CODE REVIEW

File: boa/contracts/vyper/vyper_contract.py [hunk 1]
```diff
@@ -1058,7 +1074,7 @@ def vyper_object(val, vyper_type):
     vt = type(val)
     if vt is bool:
         # https://stackoverflow.com/q/2172189
-        # bool is not ambiguous wrt vyper type anyways.
```
Comment (charles-cooper) [inline boa/contracts/vyper/vyper_contract.py:1061]: i think this is an americanism, we can leave it in :)

File: boa/contracts/vyper/vyper_contract.py [hunk 2]
```diff
@@ -1048,7 +1048,23 @@ def __init__(self, contract, fn_source):
...
+    """
+    A wrapper around str that adds a canonical_address property to mimic `Address`.
+    This is necessary because `Address` overrides `__new__`, so it always returns
+    an Address instance that cannot be modified.
+    """
+
+    def __new__(cls, *args, **kwargs):
+        self = super().__new__(cls, *args, **kwargs)
+        self.canonical_address = Address(self).canonical_address
+        return self
```
Comment (charles-cooper) [inline boa/contracts/vyper/vyper_contract.py:1064]: yea after thinking about this i don't think it's necessary. see the special handling for bool in `vyper_object()`, we can follow that for the `Address` type.

File: boa/util/abi.py [hunk 3]
```diff
@@ -1,15 +1,39 @@
...
+        value = getattr(value, "address", value)
+        return super().visit_AddressNode(node, value)
+
+
+class AbiDecoder(Decoder):
+    @classmethod
+    def visit_AddressNode(
+        cls, node: nodes.AddressNode, value: bytes, checksum: bool = True, **kwargs: Any
+    ) -> "Address":
+        from boa.environment import (  # Maybe Address should be in a different file?
```
Comment (charles-cooper) [inline boa/util/abi.py:29]: yes, either `boa.address` or move it to this file.

File: boa/util/abi.py [hunk 4]
```diff
@@ -1,15 +1,39 @@
...
 
 
+class AbiEncoder(Encoder):
+    @classmethod
+    def visit_AddressNode(cls, node: nodes.AddressNode, value) -> bytes:
+        value = getattr(value, "address", value)
+        return super().visit_AddressNode(node, value)
+
+
+class AbiDecoder(Decoder):
```
Comment (charles-cooper) [inline boa/util/abi.py:24]: `ABIDecoder`

File: boa/contracts/abi/abi_contract.py [hunk 5]
```diff
@@ -300,15 +293,16 @@ def at(self, address: Address | str) -> ABIContract:
...
     :param abi: The ABI type to parse.
-    :return: A list of types or a single type.
+    :return: The schema string for the given abi type.
     """
     if "components" in abi:
-        assert abi["type"] == "tuple"  # sanity check
-        return [_abi_from_json(item) for item in abi["components"]]
+        assert abi["type"] in ("tuple", "tuple[]")  # sanity check
+        components = [_abi_from_json(item) for item in abi["components"]]
+        return abi["type"].replace("tuple", f"({','.join(components)})")
```
Comment (charles-cooper) [inline boa/contracts/abi/abi_contract.py:305]: what is this line for?

File: boa/util/abi.py [hunk 6]
```diff
@@ -1,15 +1,66 @@
...
+        self.canonical_address = to_canonical_address(address)
+        cls._cache[address] = self
+        return self
+
+    def __repr__(self):
+        checksum_addr = super().__repr__()
+        return f"_Address({checksum_addr})"
+
+
+class ABIEncoder(Encoder):
```
Comment (charles-cooper) [inline boa/util/abi.py:48]: a line or two comment explaining what this class is for would be helpful!

File: boa/contracts/abi/abi_contract.py (hunk 5 above)
Comment (DanielSchiavini) [reply boa/contracts/abi/abi_contract.py:305]:
```python
if abi["type"] == "tuple": return f"(components)"
if abi["type"] == "tuple[]": return f"(components)[]"
return abi["type"]
```

Comment (charles-cooper) [reply boa/contracts/abi/abi_contract.py:305]:
```suggestion
        components_str = ",".join(components)
        if abi["type"] == "tuple:
            return f"({components})"
        if abi["type"] == "tuple[]":
            return f"({components})[]"
        raise ValueError("not a tuple")
```

Comment (DanielSchiavini) [reply boa/contracts/abi/abi_contract.py:305]:
```python
        match abi["type"]:
            case "tuple": return f"({components})"
            case "tuple[]": return f"({components})[]"
            case _: raise ValueError(f"Unknown type {abi['type']}")
```

Comment (charles-cooper) [reply boa/contracts/abi/abi_contract.py:305]: my only issue with this is that after linting we end up with extra indentation, but it's not really a big enough deal to be a blocker

File: boa/util/abi.py [hunk 7]
```diff
@@ -1,15 +1,75 @@
...
+from eth_typing import Address as PYEVM_Address
+from eth_utils import to_canonical_address, to_checksum_address
+
+from boa.util.lrudict import lrudict
 
 _parsers: dict[str, ABITypeNode] = {}
 
 
+# XXX: inherit from bytes directly so that we can pass it to py-evm?
+# inherit from `str` so that ABI encoder / decoder can work without failing
```
Comment (charles-cooper) [inline boa/util/abi.py:19]: oh cool, we can remove the inheritance from `str` now (and maybe replace with inheritance from bytes)

File: tests/unitary/test_abi.py [hunk 8]
```diff
@@ -89,21 +89,32 @@ def test(_a: address) -> address:
...
+    return (first, _a[2]), _a[3]
     """
     abi_contract, vyper_contract = load_via_abi(code)
     deployer_contract = abi_contract.deployer.at(abi_contract.address)
     given = [1, 2, 3, 4, 5]
     sender = Address(boa.env.eoa)
-    expected = ([sender, sender], 1)
+    expected = (([(sender, 1), (sender, 2)], 3), 4)
+    abi_result = abi_contract.test(given)
+    assert abi_result == expected
```
Comment (charles-cooper) [inline tests/unitary/test_abi.py:113]:
```suggestion

    assert abi_result == expected
```

**Review by charles-cooper — APPROVED** [review]
lgtm, thank you!

**Pull request merged**